### PR TITLE
Delegation v3

### PR DESCRIPTION
**Pull request by 0xAshish**

### CODE REVIEW

File: contracts/staking/StakeManager.sol [hunk 1]
```diff
@@ -289,24 +281,21 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
...
-    address _contract = validators[validatorId].contractAddress;
-    if (_contract != address(0x0)) {
-      amount = amount.add(ValidatorContract(_contract).withdrawRewardsValidator());
-    }
     validators[validatorId].reward = 0;
     require(token.transfer(msg.sender, amount), "Insufficent rewards");
   }
 
-  function delegationTransfer(uint256 amount, address delegator) external {
+  function delegationTransfer(uint256 amount, address delegator) external returns (bool) {
```
Comment (mankenavenkatesh) [inline contracts/staking/StakeManager.sol]: function name is confusing. pls change it to something like 'delgatorRewardWithdraw' or 'delegatorWithdrawal' etc

File: contracts/staking/DelegationManager.sol [hunk 2]
```diff
@@ -1,230 +1,274 @@
...
+          ).checkMembership(
+            accIndex,
+            accRoot[checkpointId],
+            accProof),
+      "Wrong account proof"
+      );
+
+    uint256 _rewardAmount = rewardAmount.sub(delegators[delegatorId].claimedRewards);
+    if (_rewardAmount <= slashedAmount) {
+      delegators[delegatorId].amount = delegators[delegatorId].amount.sub(slashedAmount.sub(_rewardAmount));
```
Comment (mankenavenkatesh) [inline contracts/staking/DelegationManager.sol]: only delegator amount is reduced. TotalSupply also should be reduced.

Comment (mankenavenkatesh) [reply contracts/staking/DelegationManager.sol]: Also validatorDelegation should be reduced i think

File: contracts/staking/DelegationManager.sol [hunk 3]
```diff
@@ -79,23 +86,55 @@ contract DelegationManager is IDelegationManager, ERC721Full, Lockable {
...
     require(
       delegators[delegatorId].deactivationEpoch > 0 &&
       delegators[delegatorId].deactivationEpoch.add(
       stakeManager.WITHDRAWAL_DELAY()) <= stakeManager.currentEpoch(),
-      ""
+      "Incomplete withdraw Period"
       );
+
     uint256 amount = delegators[delegatorId].amount;
     totalStaked = totalStaked.sub(amount);
```
Comment (mankenavenkatesh) [inline contracts/staking/DelegationManager.sol]: Also reduce the stake of validator in validatorDelegation

File: contracts/staking/StakeManager.sol (hunk 1 above)
Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: Thanks fixed.

File: contracts/staking/DelegationManager.sol (hunk 2 above)
Comment (0xAshish) [reply contracts/staking/DelegationManager.sol]: fixed.

File: contracts/staking/DelegationManager.sol (hunk 3 above)
Comment (0xAshish) [reply contracts/staking/DelegationManager.sol]: this is handled in `_unbond`.

File: contracts/staking/DelegationManager.sol [hunk 4]
```diff
@@ -1,230 +1,281 @@
...
 
-    ValidatorContract(validator).bond(delegatorId, delegator.amount, currentEpoch);
-    delegator.delegationStartEpoch = currentEpoch;
+    // require(delegator.lastValidatorEpoch.add(validatorHopLimit) <= currentEpoch, "Delegation_Limit_Reached");
+    Delegator storage delegator = delegators[delegatorId];
     delegator.bondedTo = validatorId;
-    stakeManager.updateValidatorState(validatorId, currentEpoch, int(delegator.amount));
-    delegator.lastValidatorEpoch = currentEpoch;
-    emit Bonding(delegatorId, validatorId, validator);
+    validatorDelegation[validatorId] = validatorDelegation[validatorId].add(delegator.amount);
```
Comment (mankenavenkatesh) [inline contracts/staking/DelegationManager.sol]: Let's say, if delegator activation epoch is not equal to current epoch, 
we are still adding delegator amount to validator.

File: contracts/staking/IDelegationManager.sol [hunk 5]
```diff
@@ -4,21 +4,38 @@ pragma solidity ^0.5.2;
 contract IDelegationManager {
   event Staked(address indexed user, uint256 indexed delegatorId, uint256 indexed activatonEpoch, uint256 amount, uint256 total);
```
Comment (mankenavenkatesh) [inline contracts/staking/IDelegationManager.sol]: Pls change event names which are colliding with StakeManager. 
On Heimdall, we search by event names irrespective of the contract from which events are triggered..

File: contracts/staking/IDelegationManager.sol [hunk 6]
```diff
@@ -4,21 +4,38 @@ pragma solidity ^0.5.2;
...
   event Unstaked(address indexed user, uint256 indexed delegatorId, uint256 amount, uint256 total);
-  event Bonding(uint256 indexed delegatorId, uint256 indexed validatorId, address indexed validatorContract);
+  event UnstakeInit(address indexed user, uint256 indexed delegatorId, uint256 indexed deactivationEpoch);
+  event Bonding(uint256 indexed delegatorId, uint256 indexed validatorId, uint256 indexed amount);
   event UnBonding(uint256 indexed delegatorId, uint256 indexed validatorId);
```
Comment (mankenavenkatesh) [inline contracts/staking/IDelegationManager.sol]: Add amount as well. Needed on heimdall

File: contracts/staking/DelegationManager.sol [hunk 7]
```diff
@@ -1,230 +1,298 @@
...
-    require(delegator.lastValidatorEpoch == 0 || delegator.lastValidatorEpoch.add(validatorHopLimit) <= currentEpoch, "Delegation_Limit_Reached");
+  function bond(uint256 delegatorId, uint256 validatorId) public onlyDelegator(delegatorId) {
+    IStakeManager stakeManager = IStakeManager(registry.getStakeManagerAddress());
+    uint256 currentEpoch = stakeManager.currentEpoch(); //TODO add 1
 
-    address validator;
-    (,,,,,,,validator,) = stakeManager.validators(validatorId);
-    require(validator != address(0x0), "Unknown validatorId or validator doesn't expect delegations");
+    if (delegators[delegatorId].bondedTo != 0) {
+      emit ReBonding(delegatorId, delegators[delegatorId].bondedTo, validatorId);
```
Comment (mankenavenkatesh) [inline contracts/staking/DelegationManager.sol]: After ReBonding event, we are also emitting Bonding Event. 
pls remove Bonding event after rebonding event.

File: contracts/staking/IDelegationManager.sol (hunk 6 above)
Comment (0xAshish) [reply contracts/staking/IDelegationManager.sol]: added thanks

File: contracts/staking/DelegationManager.sol (hunk 7 above)
Comment (0xAshish) [reply contracts/staking/DelegationManager.sol]: ignore the Rebonding event, I've kept it for the outside world. on heimdall just use bond/unbond

Comment (mankenavenkatesh) [reply contracts/staking/DelegationManager.sol]: That would be 2 transactions on heimdall which are unnecessary. 
Let's keep Rebond event

File: contracts/staking/IDelegationManager.sol (hunk 5 above)
Comment (0xAshish) [reply contracts/staking/IDelegationManager.sol]: did

File: contracts/staking/DelegationManager.sol (hunk 7 above)
Comment (0xAshish) [reply contracts/staking/DelegationManager.sol]: okey makes sense made changes

**Comment by 0xAshish** [comment]
closing it for now, since we are going with a new design.

**Pull request closed without merging**